apps/builder/app/canvas/collapsed.ts

```diff
@@ -92,8 +93,17 @@ const getInstanceSize = (instanceId: string, tagName: HtmlTags | undefined) => {
   };
 };
 
+const MAX_SIZE_TO_USE_OPTIMIZATION = 50;
+
 const recalculate = () => {
-  const instanceIds = Array.from(instanceIdSet);
+  const rootInstanceId = selectedPageStore.get()?.rootInstanceId;
+
+  const instanceIds =
+    instanceIdSet.size < MAX_SIZE_TO_USE_OPTIMIZATION
+      ? Array.from(instanceIdSet)
+      : rootInstanceId !== undefined
+      ? [rootInstanceId]
+      : [];
```

Comment:
I don't understand what we are doing here or why

Reply:
Added comment